lib/galaxy/managers/jobs.py

```diff
@@ -127,12 +127,15 @@ def index_query(
         search = payload.search
         order_by = payload.order_by
 
-        if trans.user is None:
-            # If the user is anonymous we can only return jobs for the current session history
-            if trans.galaxy_session and trans.galaxy_session.current_history_id:
-                history_id = trans.galaxy_session.current_history_id
-            else:
-                return None
+        if (
+            trans.user is None
+            and history_id is None
+            and trans.galaxy_session
+            and trans.galaxy_session.current_history_id
+        ):
+            # If the user is anonymous and no specific history_id was provided
+            # we can only return jobs from the history in the current session
+            history_id = trans.galaxy_session.current_history_id
```

Comment:
I'd suggest filtering on the session id instead, that feel more direct ? Then you can change line 226-227 to

 if history_id is None and invocation_id is None and implicit_collection_jobs_id is None:
    if trans.user:
        stmt = stmt.where(Job.user_id == trans.user.id)
    elif trans.galaxy_session:
        stmt = stmt.where(Job.session_id == trans.user.id)
    else:
        return None # Or the @nsoranzo variant to require a session ... I think we have an API decorator or depend method to annotate this on the API endpoint.